File: frontend/lib/auth/system-permissions.ts
/**
 * Quyền cấp HỆ THỐNG (tầng 1) — soi gương
 * `PMS.Application/Common/Authorization/SystemPermissions.cs` (ADR-045).
 *
 * 🔴 **Đây KHÔNG phải bản sao thứ hai của `lib/tasks/permissions.ts`.** Hai file, hai tầng,
 * không chồng lấn:
 *
 * | | Tầng 1 — file này | Tầng 2 — `lib/tasks/permissions.ts` |
 * |---|---|---|
 * | Phạm vi | Toàn hệ thống | Theo từng project |
 * | Nguồn | `EmployeeDto.permissions` (từ DB, qua claim) | `RoleInProject` của chính tôi trong project |
 * | Soi gương | `SystemPermissions.cs` | `ProjectPermissions.cs` |
 * | Đổi bằng | Màn `/admin/roles` (dữ liệu) | Đổi vai trò thành viên |
 *
 * Quyền project-scoped **không bao giờ** xuất hiện ở đây — đó là ranh giới ADR-042 dựng lên
 * và backend có test khóa (`SystemPermissionsCatalogTests`).
 */

import type { EmployeeDto } from '@/types/auth';

/**
 * Danh mục ĐÓNG. Phải khớp từng ký tự với `SystemPermissions.All` phía backend — sai một
 * chữ thì nút biến mất im lặng chứ không báo lỗi.
 */
export const SYSTEM_PERMISSIONS = {
  employeesManage: 'employees:manage',
  auditRead: 'audit:read',
  labelsManage: 'labels:manage',
  projectsCreate: 'projects:create',
  rolesManage: 'roles:manage',
} as const;

export type SystemPermission = (typeof SYSTEM_PERMISSIONS)[keyof typeof SYSTEM_PERMISSIONS];

/*
 * 🗑️ Đã gỡ `SYSTEM_PERMISSION_LABEL` (2026-08-05).
 *
 * Nó là một bảng nhãn tiếng Việt chép tay cho năm mã quyền, ghi chú là "dùng khi backend
 * chưa trả `description`" — nhưng backend **đã** trả `description` trong danh mục quyền, và
 * `/admin/roles` (nơi duy nhất cần hiển thị tên quyền) vốn đã đọc thẳng từ đó. Bảng này
 * **chưa từng có người dùng nào**: đã kiểm bằng `git grep` trên `main`.
 *
 * Đừng dựng lại. Hai bảng nhãn cho cùng một danh mục thì chắc chắn có lúc lệch, và bản chép
 * tay ở client là bản sẽ cũ đi — thêm một quyền ở backend không có gì bắt được nó ở đây.
 * Cần tên quyền thì lấy `description` từ `GET` danh mục.
 */

/**
 * Người dùng có quyền này không?
 *
 * 🔴 Fail-closed ở MỌI đường: `user` là `null`/`undefined` (chưa tải xong), hoặc
 * `permissions` không tồn tại → trả `false`.
 *
 * Nhánh `permissions` thiếu KHÔNG phải phòng xa lý thuyết: một tab đang mở lúc backend
 * được deploy vẫn giữ `employee` cũ trong store cho tới lần refresh token kế tiếp (tối đa
 * 15 phút), và bản cũ đó không có trường này. Đọc thành mảng rỗng khiến UI ẩn nút admin
 * trong 15 phút đó — đúng, vì token của họ cũng chưa có claim nên bấm vào chỉ nhận 403.
 */
export function hasPermission(
  user: EmployeeDto | null | undefined,
  permission: SystemPermission,
): boolean {
  return user?.permissions?.includes(permission) ?? false;
}

/** Có ÍT NHẤT MỘT trong các quyền — dùng để quyết định có hiện mục "Quản trị" hay không. */
export function hasAnyPermission(
  user: EmployeeDto | null | undefined,
  permissions: readonly SystemPermission[],
): boolean {
  return permissions.some((p) => hasPermission(user, p));
}

File: backend/src/PMS.Application/Common/Authorization/SystemPermissions.cs
namespace PMS.Application.Common.Authorization;

/// <summary>
/// Danh mục ĐÓNG các quyền cấp HỆ THỐNG (tầng 1) — ADR-045.
///
/// <para>
/// 🔴 <b>Luật bất di bất dịch: không mã nào ở đây được mang phạm vi project.</b> Quyền tầng 2
/// (xem/sửa task, quản lý thành viên, xem board…) đọc <c>ProjectMember.RoleInProject</c>
/// TƯƠI mỗi request qua <c>ProjectAuthorizationService</c> + <see cref="ProjectPermissions"/>,
/// và phải ở nguyên đó vì hai lý do: một người có vai trò khác nhau ở từng project (nhét vào
/// token là token phình theo số project), và PM đổi vai trò của ai đó phải có hiệu lực TỨC
/// THÌ chứ không phải sau tối đa 15 phút.
/// </para>
/// <para>
/// Cám dỗ tự nhiên của mô hình permission là thêm <c>projects:read:all</c> cho admin. Đó
/// chính là God Mode mà ADR-042 vừa đóng lại (SystemAdmin không có <b>bất kỳ</b> đặc quyền
/// nghiệp vụ nào; không phải thành viên thì nhận 404 y hệt người ngoài). Thêm nó là làm cả
/// <c>SystemAdminScopeTests</c> lẫn <c>SystemPermissionsCatalogTests</c> đỏ — cố ý.
/// </para>
/// <para>
/// <b>Vì sao <c>const string</c> chứ không phải enum:</b> <c>[Authorize(Policy = ...)]</c> cần
/// hằng biên dịch. Quan trọng hơn — mã quyền là CÙNG MỘT CHUỖI ở bốn nơi: cột
/// <c>Permissions.Code</c>, giá trị claim <c>permission</c> trong JWT, trường
/// <c>permissions[]</c> của JSON, và hằng ở <c>frontend/lib/auth/system-permissions.ts</c>.
/// Enum sẽ chèn thêm một tầng phiên dịch, tức thêm hai chỗ nữa để lệch.
/// <see cref="ProjectPermissions"/> (tầng 2) thì dùng enum + switch — khác hình dạng là cố
/// ý, nó khiến người đọc nhìn một cái là biết đang ở tầng nào.
/// </para>
/// <para>
/// <b>Thêm quyền mới cần ĐỦ BA bước:</b> thêm <c>const</c> ở đây → thêm vào <see cref="All"/>
/// → thêm vào <c>HasData</c> của <c>PermissionConfiguration</c> rồi
/// <c>dotnet ef migrations add</c>. Quên bước 3 thì <c>has-pending-model-changes</c> đỏ; quên
/// bước 2 thì <c>SystemPermissionsCatalogTests</c> đỏ. Không có đường nào lọt im lặng.
/// </para>
/// </summary>
public static class SystemPermissions
{
    /// <summary>
    /// Kiểu claim mang quyền trong JWT. <c>Program.cs</c> đặt
    /// <c>MapInboundClaims = false</c> nên chuỗi này sống nguyên vẹn tới
    /// <c>ClaimsPrincipal</c>, không bị ánh xạ sang URI dài.
    /// </summary>
    public const string ClaimType = "permission";

    /// <summary>
    /// Quản lý nhân sự: xem danh sách, khóa/mở tài khoản, đổi <c>SystemRole</c> của người khác.
    /// <para>
    /// Cố ý là MỘT mã chứ không tách <c>employees:read</c> / <c>employees:lock</c> /
    /// <c>employees:change-role</c>: <c>[Authorize]</c> hôm nay nằm ở class-level của
    /// <c>AdminEmployeesController</c> và gác cả bốn action như nhau, nên tách ra là ĐỔI HÀNH
    /// VI giấu bên trong một refactor. Tách khi có màn hình thật cần vai trò "kiểm toán viên
    /// chỉ đọc".
    /// </para>
    /// </summary>
    public const string EmployeesManage = "employees:manage";

    /// <summary>Đọc nhật ký cấp hệ thống (<c>GET /admin/audit-logs</c>).</summary>
    public const string AuditRead = "audit:read";

    /// <summary>
    /// Sửa/xóa nhãn TOÀN CỤC. Gác <c>PUT</c>/<c>DELETE /labels/{id}</c>, cố ý KHÔNG gác
    /// <c>POST /labels</c> — bất đối xứng đã có từ trước và có lý do ghi ở
    /// <c>LabelsController</c>: tạo nhãn là thao tác cộng thêm, còn xóa một nhãn gỡ chip khỏi
    /// board của MỌI project.
    /// </summary>
    public const string LabelsManage = "labels:manage";

    /// <summary>
    /// Tạo project mới. Cấp cho <b>cả</b> <c>User</c> lẫn <c>SystemAdmin</c> — thay thế policy
    /// <c>can-create-project</c> cũ vốn chỉ là <c>RequireAuthenticatedUser()</c>, tức một
    /// no-op. Giữ nguyên hành vi: mọi người dùng đều tạo được project và tự thành PM của nó.
    /// <para>
    /// 🔴 Đây là mã <c>projects:*</c> DUY NHẤT được phép tồn tại, và nó KHÔNG vi phạm luật
    /// "không có quyền project-scoped": lúc gọi endpoint này <b>chưa có project nào</b> để tra
    /// membership, nên không có tầng 2 để đi qua. Mọi động từ project khác đều cần một project
    /// đã tồn tại → đi qua <c>ProjectAuthorizationService</c>.
    /// <c>SystemPermissionsCatalogTests</c> gọi đích danh ngoại lệ này, nên
    /// <c>projects:read</c> hay <c>projects:read:all</c> vẫn đỏ ngay.
    /// </para>
    /// </summary>
    public const string ProjectsCreate = "projects:create";

    /// <summary>
    /// Sửa ánh xạ vai trò → quyền (<c>/admin/roles</c>).
    /// <para>
    /// Đây là quyền TỰ PHỤC HỒI duy nhất: có nó thì cấp lại được mọi quyền khác qua UI. Vì
    /// vậy <c>RolePermissionAdminService</c> có bất biến "SystemAdmin luôn giữ
    /// <c>roles:manage</c>" (409) — gỡ nó là khóa vĩnh viễn mọi lối vào quản trị, vì
    /// <c>DbSeeder</c> không chạy ở production và <c>HasData</c> chỉ áp lúc migrate mới.
    /// </para>
    /// </summary>
    public const string RolesManage = "roles:manage";

    /// <summary>
    /// Toàn bộ danh mục. Dùng để đăng ký policy (<c>Program.cs</c>), để validate đầu vào của
    /// API phân quyền, và làm mốc so cho test khóa danh mục.
    /// </summary>
    public static readonly IReadOnlyList<string> All =
    [
        EmployeesManage,
        AuditRead,
        LabelsManage,
        ProjectsCreate,
        RolesManage
    ];
}
